### PR TITLE
feat: SpawnComponent

### PR DESCRIPTION
**Pull request by spydon**
# Description
This PR introduces the `SpawnComponent`, which randomly spawns components within a set area.

## Checklist

- [x] I have followed the [Contributor Guide] when preparing my PR.
- [x] I have updated/added tests for ALL new/updated/fixed functionality.
- [x] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- [x] I have updated/added relevant examples in `examples` or `docs`.

## Breaking Change?

- [ ] Yes, this PR is a breaking change.
- [x] No, this PR is not a breaking change.

[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Conventional Commit]: https://conventionalcommits.org
[CHANGELOG]: https://github.com/flame-engine/flame/blob/main/CHANGELOG.md

### CODE REVIEW

File: packages/flame/lib/src/components/core/component.dart [hunk 1]
```diff
@@ -302,7 +302,7 @@ class Component {
   ///
   /// If [includeSelf] is set to true (default is false) then the component
   /// which the call is made for is also included in the search.
-  T? findParent<T extends Component>({bool includeSelf = false}) {
```
Comment (luanpotter) [inline packages/flame/lib/src/components/core/component.dart:305]: how can your ancestor not be a component?

File: packages/flame/lib/src/experimental/geometry/shapes/polygon.dart [hunk 2]
```diff
@@ -241,4 +243,35 @@ class Polygon extends Shape {
...
+        final randomX = min.x + randomGenerator.nextDouble() * (max.x - min.x);
+        final randomY = min.y + randomGenerator.nextDouble() * (max.y - min.y);
+        result.setValues(randomX, randomY);
+
+        if (containsPoint(result)) {
+          return result;
+        }
+      }
+    } else {
+      final edgeIndex = randomGenerator.nextInt(vertices.length);
```
Comment (luanpotter) [inline packages/flame/lib/src/experimental/geometry/shapes/polygon.dart:265]: not uniformly distributed bc the edges can have different lengths and thus should have different weights.
how about adding up the perimeter, generate a random number and find the edge and then find the point

File: packages/flame/lib/src/experimental/geometry/shapes/polygon.dart [hunk 3]
```diff
@@ -241,4 +243,35 @@ class Polygon extends Shape {
...
+
+  @override
+  Vector2 randomPoint({Random? random, bool within = true}) {
+    final randomGenerator = random ?? randomFallback;
+    if (within) {
+      final result = Vector2.zero();
+      final min = aabb.min;
+      final max = aabb.max;
+
+      while (true) {
```
Comment (luanpotter) [inline packages/flame/lib/src/experimental/geometry/shapes/polygon.dart:255]: I think there are ways to do this, at least for convex polygons
https://blogs.sas.com/content/iml/2020/10/21/random-points-in-polygon.html
but this is probably good enough, I wouldn't expect it to be slow for "normal" shapes

File: packages/flame/lib/src/experimental/geometry/shapes/rectangle.dart [hunk 4]
```diff
@@ -161,6 +162,44 @@ class Rectangle extends Shape {
...
+  @override
+  Vector2 randomPoint({Random? random, bool within = true}) {
+    final randomGenerator = random ?? randomFallback;
+    if (within) {
+      return Vector2(
+        left + randomGenerator.nextDouble() * width,
+        top + randomGenerator.nextDouble() * height,
+      );
+    } else {
+      final edge = randomGenerator.nextInt(4);
```
Comment (luanpotter) [inline packages/flame/lib/src/experimental/geometry/shapes/rectangle.dart:174]: same issue

File: packages/flame/lib/src/experimental/geometry/shapes/rounded_rectangle.dart [hunk 5]
```diff
@@ -197,7 +197,9 @@ class RoundedRectangle extends Shape {
   @override
   String toString() =>
       'RoundedRectangle([$_left, $_top], [$_right, $_bottom], $_radius)';
-}
 
-@internal
-const tau = Transform2D.tau; // 2π
+  @override
+  Vector2 randomPoint({Random? random, bool within = true}) {
+    throw UnimplementedError();
```
Comment (luanpotter) [inline packages/flame/lib/src/experimental/geometry/shapes/rounded_rectangle.dart:203]: you could use the rectangle algo and do the contains check on a while loop, like the arbitrary poly

File: packages/flame/lib/src/components/core/component.dart (hunk 1 above)
Comment (spydon) [reply packages/flame/lib/src/components/core/component.dart:305]: If you check against a mixin for example, which I think is a pretty common usecase

File: packages/flame/lib/src/experimental/geometry/shapes/polygon.dart (hunk 2 above)
Comment (spydon) [reply packages/flame/lib/src/experimental/geometry/shapes/polygon.dart:265]: True true! That will be quite expensive though, I guess if we use length2 it shouldn't be a problem.

File: packages/flame/lib/src/components/core/component.dart (hunk 1 above)
Comment (luanpotter) [reply packages/flame/lib/src/components/core/component.dart:305]: wouldn't the mixin be `on Component` or `on PositionComponent`?

File: packages/flame/lib/src/experimental/geometry/shapes/polygon.dart (hunk 3 above)
Comment (spydon) [reply packages/flame/lib/src/experimental/geometry/shapes/polygon.dart:255]: Yeah, I wanted it to work for concave ones too, and there doesn't seem to be any good efficient algorithm for that.

File: packages/flame/lib/src/components/core/component.dart (hunk 1 above)
Comment (spydon) [reply packages/flame/lib/src/components/core/component.dart:305]: Not necessarily, see for example `PositionProvider` and all the other "provider" classes in that file,

File: packages/flame/lib/src/components/spawn_component.dart [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    this.area,
+    this.within = true,
+    double? period,
+    double? minPeriod,
+    double? maxPeriod,
+    Random? random,
+    super.key,
+  })  : _period = period ?? 0,
+        _random = random ?? _randomFallback,
+        assert(
```
[the remark this answers is not in the document]
Comment (spydon) [reply packages/flame/lib/src/components/spawn_component.dart:27]: How do you mean? They either have to specify `period` or both `minPeriod` and `maxPeriod`, so all of them can't be required?

File: packages/flame/lib/src/components/core/component.dart (hunk 1 above)
Comment (spydon) [reply packages/flame/lib/src/components/core/component.dart:305]: On the other hand, then that type is not really a `parent`, so I guess it is better to go through the `ancestors` method instead.

File: doc/flame/components.md [hunk 7]
```diff
@@ -340,6 +340,12 @@ void onDragUpdate(DragUpdateInfo info) {
 
 ### PositionType
 
+```{note}
+If you are using the `CameraComponent` you should not use `PositionType`, but
```
Comment (erickzanardo) [inline doc/flame/components.md:344]: Is this related to this new component?

File: packages/flame/lib/src/components/spawn_component.dart [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+          period != null || (minPeriod != null && maxPeriod != null),
+          'Either the static period must be defined, or both minPeriod and '
+          'maxPeriod must be defined',
+        );
+
+  /// The function used to create new components to spawn.
+  PositionComponent Function() factory;
```
Comment (erickzanardo) [inline packages/flame/lib/src/components/spawn_component.dart:34]: maybe would be cool to receive the index of the created component, maybe people could do some logic on top of it, like, spawn X is even and Y if odd, or every 10 spawnmed, spawn a boss.

File: doc/flame/components.md (hunk 7 above)
Comment (spydon) [reply doc/flame/components.md:344]: It is not, I just saw that the docs missed it since I added this component to the section underneath.

File: packages/flame/lib/src/components/spawn_component.dart (hunk 8 above)
Comment (spydon) [reply packages/flame/lib/src/components/spawn_component.dart:34]: It loses the nice way of being able to do `factory: MyComponent.new`, but it is true that the index could be useful, so I'll add it!

File: doc/flame/components.md (hunk 7 above)
Comment (erickzanardo) [reply doc/flame/components.md:344]: Got ya!

File: packages/flame/lib/src/components/spawn_component.dart (hunk 8 above)
Comment (erickzanardo) [reply packages/flame/lib/src/components/spawn_component.dart:34]: yeah, it is a shame that we don't have unions yet :(

File: packages/flame/lib/src/components/spawn_component.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (luanpotter) [reply packages/flame/lib/src/components/spawn_component.dart:27]: that is why I was thinking of two named constructors, `.period()` and `.random()` (or something), so for each the parameters for that type of setup are clearly defined in a type safe way.
the can both all call this, private version where everything is nullable and the check is at runtime

File: packages/flame/lib/src/experimental/geometry/shapes/rounded_rectangle.dart (hunk 5 above)
Comment (spydon) [reply packages/flame/lib/src/experimental/geometry/shapes/rounded_rectangle.dart]: Done! It won't work for `within = false` though.

File: packages/flame/lib/src/components/spawn_component.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (spydon) [reply packages/flame/lib/src/components/spawn_component.dart]: Aaah of course, I read named arguments in my head for some reason. Fixed!

File: packages/flame/lib/src/components/spawn_component.dart (hunk 8 above)
Comment (spydon) [reply packages/flame/lib/src/components/spawn_component.dart]: For this use-case unions would be really dirty since it then would have to check what the function signature is in runtime to be able to either send or not send the argument

**Review by erickzanardo — APPROVED** [review]

File: packages/flame/lib/src/experimental/geometry/shapes/polygon.dart [hunk 9]
```diff
@@ -241,4 +243,58 @@ class Polygon extends Shape {
...
+    while (vertexIndex < verticesLengths2.length) {
+      currentEndPoint += verticesLengths2[vertexIndex];
+      if (currentEndPoint >= pointOnEdges) {
+        break;
+      }
+      vertexIndex++;
+    }
+    final startPoint = vertices[vertexIndex];
+    final endPoint = vertices[(vertexIndex + 1) % vertices.length];
+    final t = randomGenerator.nextDouble();
```
Comment (luanpotter) [inline packages/flame/lib/src/experimental/geometry/shapes/polygon.dart:296]: you don't need to generate another random double. just decrease pointOnEdges as you iterate the edges and t is the "remainder" for this edge

Comment (spydon) [reply packages/flame/lib/src/experimental/geometry/shapes/polygon.dart:296]: Ah, good idea!

File: packages/flame/lib/src/math/random_fallback.dart [hunk 10]
```diff
@@ -0,0 +1,5 @@
+import 'dart:math';
+
+/// When you don't care about what [Random] object you have and don't want to
+/// create an unnecessary object you can use this pre-created object.
+final Random randomFallback = Random();
```
Comment (luanpotter) [inline packages/flame/lib/src/math/random_fallback.dart:5]: is this just to avoid instantiating a random? I feel `?? Random()` is actually shorter than `?? randomFallback`

**Review by luanpotter — APPROVED** [review]
LGTM. Thanks for unifying tau, it has been bothering me!

File: packages/flame/lib/src/math/random_fallback.dart (hunk 10 above)
Comment (spydon) [reply packages/flame/lib/src/math/random_fallback.dart:5]: Exactly, so that there aren't any new `Random` objects created in tight loops

File: packages/flame/lib/src/experimental/geometry/shapes/polygon.dart (hunk 9 above)
Comment (spydon) [reply packages/flame/lib/src/experimental/geometry/shapes/polygon.dart:296]: Hmm, then we'll have to do a square root though, I think it probably is cheaper to generate another double?
